### PR TITLE
[Mobile] - RichText - Use parsed font size values when comparing new changes

### PR DESCRIPTION
**Pull request by geriux**
- Gutenberg Mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/4463

## Description
This PR fixes a bug within `RichText`, where in some cases it would generate a render loop when setting a new font-size value and this didn't match the parsed value.

With this change it'll parse the `currentFontSizeStyle` and `prevFontSizeStyle` before comparing values to update the state. As well as **not using** the font size value from its props if there's a font size provided from the styles.

**Note:** The font size picker is behind the `__DEV__` flag so it'll only show up with metro running for block-based themes. But **it will render** the font-size value correctly for both default and block-based themes.

### Test case 1
- Open the app
- Switch to the HTML mode
- Set this content

```
<!-- wp:paragraph {"style":{"color":{"text":"#fcb900"},"typography":{"fontSize":35.56}}} -->
<p class="has-text-color" style="color:#fcb900;font-size:35.56px">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed imperdiet ut nibh vitae ornare. Sed auctor nec augue at blandit.</p>
<!-- /wp:paragraph -->
```
- Go back to the visual mode
- Expect the `Paragraph` block to show correctly.

### Test case 2

**Precondition**: Using a block-based theme.

- Open the WordPress app
- Start a new post
- Within the `Paragraph` block type some text
- Open the block settings
- Change the font size using the font size picker
- **Expect** to see the new selected font size.

### Test case 3

**Precondition**: Using a block-based theme.

- Open the WordPress app
- Start a new post
- Add a `Heading` block and type some text
- Change between different heading levels and **expect** the size to change correctly
- Change the font size using the font picker and **expect** the size to change correctly

## Screenshots 

Test case 1 | Test case 2 | Test case 3
-|-|-
<kbd></kbd> | <kbd></kbd> | <kbd></kbd>

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [x] I've updated related schemas if appropriate.

### CODE REVIEW

**Review by fluiddot — COMMENTED** [review]
I've added locally a test case to the integration test for the `RichText` component ([reference](https://github.com/WordPress/gutenberg/blob/trunk/packages/rich-text/src/test/index.native.js)) for checking the issue and I managed to reproduce the infinite loop issue:
```
it( 'should update the font size with provided style', () => {
	// Arrange
	const fontSize = '10';
	const style = { fontSize: '12' };
	// Act
	const screen = render( <RichText fontSize={ fontSize } /> );
	screen.update(
		<RichText fontSize={ fontSize } style={ style } />
	);
	// Assert
	expect( screen.toJSON() ).toMatchSnapshot();
} );
```

I tested a different flow where the font size could be updated when the `style` prop with a `fontSize` property is passed once the component is already mounted, not sure if this case is possible but I bumped into the same loop issue:

```
Maximum update depth exceeded. This can happen when a component repeatedly calls setState inside componentWillUpdate or componentDidUpdate. React limits the number of nested updates to prevent infinite loops.
```

**Comment by geriux** [comment]
> I tested a different flow where the font size could be updated when the `style` prop with a `fontSize` property is passed once the component is already mounted, not sure if this case is possible but I bumped into the same loop issue:

Thank you @fluiddot! I'll check it out! 🙇

**Comment by fluiddot** [comment]
> > I tested a different flow where the font size could be updated when the `style` prop with a `fontSize` property is passed once the component is already mounted, not sure if this case is possible but I bumped into the same loop issue:
> 
> Thank you @fluiddot! I'll check it out! 🙇

I was also wondering if we could cover the test case you described checking the following HTML via an integration test:
```
<!-- wp:paragraph {"style":{"color":{"text":"#fcb900"},"typography":{"fontSize":35.56}}} -->
<p class="has-text-color" style="color:#fcb900;font-size:35.56px">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed imperdiet ut nibh vitae ornare. Sed auctor nec augue at blandit.</p>
<!-- /wp:paragraph -->
```

I could help with that in case you think it's worth it 🙇 .

**Comment by fluiddot** [comment]
> > > I tested a different flow where the font size could be updated when the `style` prop with a `fontSize` property is passed once the component is already mounted, not sure if this case is possible but I bumped into the same loop issue:
> > 
> > 
> > Thank you @fluiddot! I'll check it out! 🙇
> 
> I was also wondering if we could cover the test case you described checking the following HTML via an integration test:
> 
> ```
> 
> <p class="has-text-color" style="color:#fcb900;font-size:35.56px">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed imperdiet ut nibh vitae ornare. Sed auctor nec augue at blandit.</p>
> 
> ```
> 
> I could help with that in case you think it's worth it 🙇 .

In case it helps, here is a diff patch that adds a couple of test cases:
<details><summary>Click here to show diff patch</summary>

```diff
diff --git a/packages/rich-text/src/test/__snapshots__/index.native.js.snap b/packages/rich-text/src/test/__snapshots__/index.native.js.snap
new file mode 100644
index 0000000000000000000000000000000000000000..eb684a9a01299ba4eb9f435fd81bb6654b8c6b0a
--- /dev/null
+++ b/packages/rich-text/src/test/__snapshots__/index.native.js.snap
@@ -0,0 +1,7 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`<RichText/> Font Size renders component with style and font size 1`] = `
+"<!-- wp:paragraph {\\"style\\":{\\"color\\":{\\"text\\":\\"#fcb900\\"},\\"typography\\":{\\"fontSize\\":35.56}}} -->
+<p class=\\"has-text-color\\" style=\\"color:#fcb900;font-size:35.56px\\">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed imperdiet ut nibh vitae ornare. Sed auctor nec augue at blandit.</p>
+<!-- /wp:paragraph -->"
+`;
diff --git a/packages/rich-text/src/test/index.native.js b/packages/rich-text/src/test/index.native.js
index b48d8828326494e3acebd8a944cc6471f4dfc030..a6ce5a233fb3e55d4ca1a911934a2867834eee66 100644
--- a/packages/rich-text/src/test/index.native.js
+++ b/packages/rich-text/src/test/index.native.js
@@ -1,17 +1,27 @@
-jest.mock( '@wordpress/data/src/components/use-select' );
 /**
  * External dependencies
  */
 import { Dimensions } from 'react-native';
-import { render } from 'test/helpers';
+import { getEditorHtml, render, initializeEditor } from 'test/helpers';
+
 /**
  * WordPress dependencies
  */
-import { useSelect } from '@wordpress/data';
+import { select } from '@wordpress/data';
+import { store as blockEditorStore } from '@wordpress/block-editor';
+import { coreBlocks, registerBlock } from '@wordpress/block-library';
+import {
+	getBlockTypes,
+	setDefaultBlockName,
+	unregisterBlockType,
+} from '@wordpress/blocks';
+import '@wordpress/jest-console';
+
 /**
  * Internal dependencies
  */
 import RichText from '../component/index.native';
+import { store as richTextStore } from '../store';
 
 /**
  * Mock `useSelect` with various global application settings, e.g., styles.
@@ -25,16 +35,13 @@ const mockGlobalSettings = (
 	const DEFAULT_GLOBAL_STYLES = {
 		__experimentalGlobalStylesBaseStyles: { typography: { fontSize } },
 	};
-	const selectMock = {
-		getFormatTypes: jest.fn().mockReturnValue( [] ),
-		getBlockParents: jest.fn(),
-		getBlock: jest.fn(),
-		getSettings: jest.fn().mockReturnValue( DEFAULT_GLOBAL_STYLES ),
-	};
 
-	useSelect.mockImplementation( ( callback ) => {
-		return callback( () => selectMock );
-	} );
+	jest.spyOn( select( blockEditorStore ), 'getSettings' ).mockReturnValue(
+		DEFAULT_GLOBAL_STYLES
+	);
+	jest.spyOn( select( richTextStore ), 'getFormatTypes' ).mockReturnValue(
+		[]
+	);
 };
 
 describe( '<RichText/>', () => {
@@ -51,6 +58,13 @@ describe( '<RichText/>', () => {
 		[ '1.42vh', 19 ],
 	];
 
+	beforeAll( () => {
+		// Register Paragraph block
+		const paragraph = coreBlocks[ 'core/paragraph' ];
+		registerBlock( paragraph );
+		setDefaultBlockName( paragraph.name );
+	} );
+
 	beforeEach( () => {
 		mockGlobalSettings( {} );
 	} );
@@ -59,6 +73,13 @@ describe( '<RichText/>', () => {
 		Dimensions.set( { window } );
 	} );
 
+	afterAll( () => {
+		// Clean up registered blocks
+		getBlockTypes().forEach( ( block ) => {
+			unregisterBlockType( block.name );
+		} );
+	} );
+
 	describe( 'Font Size', () => {
 		it( 'should display rich text at the DEFAULT font size.', () => {
 			// Arrange
@@ -193,5 +214,27 @@ describe( '<RichText/>', () => {
 			const actualFontSize = getByA11yLabel( 'editor' ).props.fontSize;
 			expect( actualFontSize ).toBe( expectedFontSize );
 		} );
+
+		it( 'should update the font size when style prop with font size property is provided', () => {
+			// Arrange
+			const fontSize = '10';
+			const style = { fontSize: '12' };
+			// Act
+			const screen = render( <RichText fontSize={ fontSize } /> );
+			screen.update( <RichText fontSize={ fontSize } style={ style } /> );
+			// Assert
+			expect( screen.toJSON() ).toMatchSnapshot();
+		} );
+
+		it( 'renders component with style and font size', async () => {
+			// Arrange
+			const initialHtml = `<!-- wp:paragraph {"style":{"color":{"text":"#fcb900"},"typography":{"fontSize":35.56}}} -->
+			<p class="has-text-color" style="color:#fcb900;font-size:35.56px">Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed imperdiet ut nibh vitae ornare. Sed auctor nec augue at blandit.</p>
+			<!-- /wp:paragraph -->`;
+			// Act
+			await initializeEditor( { initialHtml } );
+			// Assert
+			expect( getEditorHtml() ).toMatchSnapshot();
+		} );
 	} );
 } );
```

</details>

Note that I changed the way the store selectors are mocked, instead of mocking the entire `useSelect` I only mocked the selectors used in the component. This is required for rendering the editor, as it uses several selectors which doesn't make sense to mock them in the test only for this purpose.

**Comment by geriux** [comment]
> In case it helps, here is a diff patch that adds a couple of test cases:

Awesome! Thanks for the diff Carlos, I'll apply that and remove the E2E test 👍

**Comment by geriux** [comment]
@fluiddot Thank you for the suggestions, I've updated the code and PR descripton. This is ready for another review whenever you have some time 🙇

File: packages/rich-text/src/test/__snapshots__/index.native.js.snap [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    accessibilityLabel="Text input. Empty"
+    activeFormats={Array []}
+    blockType={
+      Object {
+        "tag": "div",
+      }
+    }
+    disableEditingMenu={false}
+    fontFamily="serif"
+    fontSize={13}
```
Comment (fluiddot) [inline packages/rich-text/src/test/__snapshots__/index.native.js.snap:50]: I'm wondering if the value should be rounded, as the provided font-size value is `12.56px` ([reference](https://github.com/WordPress/gutenberg/blob/60ca367fba5b2ea3d976151cb17746eec243de4f/packages/rich-text/src/test/index.native.js#L242)) 🤔 . I tested this behavior on the web version by setting the value manually via the block settings and the HTML conserved the decimals, @geriux do you know what's the expected behavior in this case?


Screenshot 2022-01-14 at 14 24 47

**HTML output:**
```
<!-- wp:paragraph {"style":{"typography":{"fontSize":"12.56px"}}} -->
<p style="font-size:12.56px">Hello world!</p>
<!-- /wp:paragraph -->
```

**Review by fluiddot — APPROVED** [review]
LGTM 🎊 

I performed tests on both platforms for covering the three described tests cases, all of them passed ✅ .

However, I spotted a couple of issues when testing on Android that I think they're not related to this change, as it works on iOS, but that are worth it to share:

### Text overlaps on H1 headings
**How to reproduce it:**
1. Add a Heading block.
2. Set H1 level.
3. Type a text long enough for being displayed in two or more lines.
4. Observe that part of the text is overlapped.



### Specified font size is not preserved when changing heading level
I verified that the behavior on both web and iOS when setting a font size is that changing the heading level doesn't modify the size. However, on Android, this is working differently (see attached video capture).

https://user-images.githubusercontent.com/14905380/149524014-5f2a378e-52e9-4721-a9d2-6326c17e7214.mp4

_Tested on Samsung Galaxy S20 FE 5G (Android 10) and Simulator - iPhone 12 Pro Max (iOS 15.0)._

File: packages/rich-text/src/test/__snapshots__/index.native.js.snap (hunk 1 above)
Comment (geriux) [reply packages/rich-text/src/test/__snapshots__/index.native.js.snap:50]: Currently, we are rounding the value yes. We could change this behavior when we add support to add custom font sizes. Thanks for double-checking!

**Comment by geriux** [comment]
Thank you for testing and reviewing! 🎉 

I'll investigate those Android issues regarding the line height and the Heading block to make sure those are fixed before releasing it to production 👍

**Pull request merged**